WordPress/gutenberg
# [RNMobile] Scroll to selected block that goes off the screen

Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/735

Gutenberg Mobile PR https://github.com/wordpress-mobile/gutenberg-mobile/pull/2771
## Description

This solution seeks to address the issue where a block may get cut off due to it being moved off the edge of the screen. In a nutshell, whenever the selected `BlockListItem` component (which is a parent component for rendered blocks) is updated then we check to see if this list item is within the current viewport and if not we are utilizing the underlying scroll behavior for the respective Lists on each platform to reposition that item on the screen.  

## How has this been tested?
To test follow the instructions below. 

1. Open a post. 
2. Select a block and press the **move** down button until the block reaches the edge of the screen. 
3. Observe that once the **Selected** block that is being moved is positioned outside the bounds of the screen then the editor automatically scrolls to that position. 
4. If you select a block and press the **move** up button, then the same behavior as expounded on above will take place. The editor will scroll instantaneously so that the block is in the viewport. 

## Screenshots 
Below are two gifs of the functionality in action on Android and iOS

### Moving a block off screen will auto-scroll the editor on Android

<kbd><img src="https://user-images.githubusercontent.com/1509205/97938345-309ab800-1d4f-11eb-98af-9a4ca895c5d8.gif" width="320"></kbd>

### Moving a block off screen will auto-scroll the editor on iOS

<kbd><img src="https://user-images.githubusercontent.com/1509205/97938354-36909900-1d4f-11eb-9eeb-50b62917b8bc.gif" width="320"></kbd>

## Types of changes

The main tenets of the solution are as follows : 

- The `measure` method of the `View` component is being utilized to determine the width, height, and position of the view within the current page/screen. 
- The `measureLayout` method of the `View` component detects where within the `FlatList` the current `BlockListItem` exists so that we can utilize it's `y` position for scrolling behavior. 
- For scrolls, the implementation differs per platform because the `FlatList`'s scroll functionality on iOS is managed by the `KeyboardAwareFlatList` so the ref to it's underlying `ScrollView` was utilized while on Android, scrolling was accomplished using the `FlatList` component itself.  
- To detect if a `BlockListItem` is within the window/screen we check to see if the `pageY` value is below zero ( which means we would need to scroll up) or above the window height ( which means we would need to scroll down)
- You will notice within the code that we are utilizing `scrollAdjustmentY` in the scroll algorithm. This is to give the `BlockListItem` the ability to initiate scrolling without it's bottom or top being at the edge of the screen. This is also used within the calculation of the offset to scroll to so that we don't have an issue where we scroll to the bottom or top of a view and there's no space above or below the block.  

## Concerns

1. On Android, I have noticed that scrolling a large block will sometimes cause a bit of the block to be cut off. I will need to have another look at the algorithm in place to ensure that it incorporates the height of the block in the right way so this never happens. 
2. I realized that scrolling a block pass a really large block that takes up the entire screen messes with its scrolling. It's like it stops scrolling on that very page because the block was moved a page away. 

Below you will observe that the list scrolls but it stops on the screen with the block that has taken up the entire viewport. 

<kbd><img src="https://user-images.githubusercontent.com/1509205/97938887-976ca100-1d50-11eb-8070-10d01c645a69.gif" width="320"></kbd>

You will also observe the same behavior when trying to scroll up with the block up movement action. The block is not scrolled to. The list stops the scroll action on the part of the page with the larger block. 

<kbd><img src="https://user-images.githubusercontent.com/1509205/97938893-9b98be80-1d50-11eb-993f-0cc1c07fec9f.gif" width="320"></kbd>

I am wondering if this is being caused because a `FlatList` is `Virtualized` so the scroll is faster than the fill rate of the content being rendered asynchronously off-screen. If so, then probably a delay could be introduced to ensure the list has rendered or we would have to somehow extend the off-screen limit in some way. Another approach would be to have some form of timeout/flag mechanism that runs a check to see if the block is within the viewport after the scroll has completed and if not then it does another scroll to the position it was supposed to scroll to in the first place.

## Checklist:
